src/app/ReadClient.cpp

```diff
@@ -0,0 +1,249 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines read client for a CHIP Interaction Data model
+ *
+ */
+
+#include <app/InteractionModelEngine.h>
+#include <app/ReadClient.h>
+
+namespace chip {
+namespace app {
+
+CHIP_ERROR ReadClient::Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate)
+{
+    CHIP_ERROR err = CHIP_NO_ERROR;
+    // Error if already initialized.
+    VerifyOrExit(mpExchangeMgr == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrExit(mpExchangeCtx == nullptr, err = CHIP_ERROR_INCORRECT_STATE);
+
+    mpExchangeMgr = apExchangeMgr;
+    mpExchangeCtx = nullptr;
+    mpDelegate    = apDelegate;
+    mState        = ClientState::Initialized;
+
+exit:
+    ChipLogFunctError(err);
+    return err;
+}
+
+void ReadClient::Shutdown()
+{
+    ClearExistingExchangeContext();
+    mpExchangeMgr = nullptr;
+    mpDelegate    = nullptr;
+    MoveToState(ClientState::Uninitialized);
+}
+
+void ReadClient::Reset()
+{
+    ClearExistingExchangeContext();
+    MoveToState(ClientState::Initialized);
+}
+
+const char * ReadClient::GetStateStr() const
+{
+#if CHIP_DETAIL_LOGGING
+    switch (mState)
+    {
+    case ClientState::Uninitialized:
+        return "UNINIT";
+    case ClientState::Initialized:
+        return "INIT";
+    case ClientState::ReadRequestSending:
+        return "READREQUESTSENDING";
+    }
+#endif // CHIP_DETAIL_LOGGING
+    return "N/A";
+}
+
+void ReadClient::MoveToState(const ClientState aTargetState)
+{
+    mState = aTargetState;
+    ChipLogDetail(DataManagement, "Client[%u] moving to [%5.5s]",
+                  InteractionModelEngine::GetInstance()->GetReadClientArrayIndex(this), GetStateStr());
+}
+
+void ReadClient::ClearState(void)
+{
+    MoveToState(ClientState::Uninitialized);
+}
+
+CHIP_ERROR ReadClient::SendReadRequest(NodeId aNodeId, Transport::AdminId aAdminId)
+{
+    CHIP_ERROR err                        = CHIP_NO_ERROR;
+    System::PacketBufferHandle msgBuf     = nullptr;
+    EventPathParams * eventPathParamsList = nullptr;
+    size_t aEventPathParamsListSize       = 0;
+    chip::EventNumber eventNumber         = 0;
+    ChipLogDetail(DataManagement, "Client[%u] [%5.5s]", InteractionModelEngine::GetInstance()->GetReadClientArrayIndex(this),
```

Comment:
Should the string mention something about us being in SendReadRequest?

Reply:
Done, thanks